File: 2/checksum.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func part1() {
	f, err := os.Open("./input/input.txt")
	defer f.Close()
	if err != nil {
		panic(err)
	}

	scanner := bufio.NewScanner(f)
	checksum := 0

	for scanner.Scan() {
		pieces := strings.Split(scanner.Text(), "\t")
		var min int = -1
		var max int = -1

		nums := sstois(pieces)

		for _, num := range nums {
			if min == -1 || num < min {
				min = num
			}
			if max == -1 || num > max {
				max = num
			}
		}
		checksum += (max - min)
	}

	fmt.Printf("Part 1: %d\n", checksum)

	if err := scanner.Err(); err != nil {
		panic(err)
	}
}

func part2() {
	f, err := os.Open("./input/input.txt")
	defer f.Close()
	if err != nil {
		panic(err)
	}

	scanner := bufio.NewScanner(f)
	checksum := 0

	for scanner.Scan() {
		pieces := strings.Split(scanner.Text(), "\t")
		is := sstois(pieces)

		length := len(is)
	RowLoop:
		for i := 0; i < length-1; i++ {
			for j := i + 1; j < length; j++ {
				n := float64(is[i])
				d := float64(is[j])
				if d > n {
					tmp := n
					n = d
					d = tmp
				}

				if math.Remainder(n, d) == 0.0 {
					checksum += int(n / d)
					break RowLoop
				}
			}
		}
	}

	fmt.Printf("Part 2: %d\n", checksum)

	if err := scanner.Err(); err != nil {
		panic(err)
	}
}

func sstois(ss []string) []int {
	is := make([]int, len(ss))
	for i, v := range ss {
		num, err := strconv.Atoi(v)
		if err != nil {
			panic(err)
		}
		is[i] = num
	}

	return is
}

func main() {
	part1()
	part2()
}

File: 2/checksum.rs
use std::fs::File;
use std::io::Read;

fn part1() {
	let input = get_input();
	let mut sum = 0;

	for line in input.trim().split("\n") {
		let nums = get_nums(line.to_string());
		let mut min = 0;
		let mut max = 0;

		for num in nums {
			if min == 0 || num < min {
				min = num;
			}
			if max == 0 || num > max {
				max = num;
			}
		}

		sum += max - min;
	}

	println!("Part 1: {}", sum);
}

fn part2() {
	let input = get_input();
	let mut sum = 0;

	for line in input.trim().split("\n") {
		let nums = get_nums(line.to_string());
		let size = nums.len();
		for i in 0..size {
			for j in (i + 1)..size {
				let mut n = nums[i] as f64;
				let mut d = nums[j] as f64;
				if n < d {
					let tmp = n;
					n = d;
					d = tmp;
				}

				let div = n / d;
				if div.fract() == 0.0 {
					sum += div as u32;
				}
			}
		}
	}

	println!("Part 2: {}", sum);
}

fn get_nums(s: String) -> Vec<u32> {
	s.trim().split("\t").map(|s| { s.parse::<u32>().unwrap() }).collect()
}

fn get_input() -> String {
	let mut file = File::open("./input/input.txt").unwrap();
	let mut contents = String::new();
	file.read_to_string(&mut contents).unwrap();

	return contents;
}

fn main() {
	part1();
	part2();
}
